### DISCUSSION TITLE
How to improve the performance of the application

### DISCUSSION THREAD

**Discussion by nuwanb-swivel** [handsontable/hyperformula#1241] [General]
We have an application with around 5000 records on the ‘aggregation’ sheet. We also have 8 more sheets that contain some codes, each defined as named expressions. These codes are used on the aggregation sheet. These codes will be used inside the aggregation sheet on different columns as a text or inside the formula. We need to calculate the usage count for each code on each sheet and also calculate the total if the code is used in the aggregation sheet.

<img width="544" alt="Screenshot 2023-04-15 at 7 44 38 AM" src="https://user-images.githubusercontent.com/114735741/232178493-b8b705f5-e572-4f55-9c93-039da702f18b.png">

**Formulas**
Total
`=IF(groups!C1="","0",IF(I1="N",SUMIF(aggregations!B:B,"*" & groups!C1& "*",aggregations!K:K),SUMIF(aggregations!B:B,"*" & groups!C1& "*",aggregations!M:M)))`

Usage count
`=IF(groups!C1="","0",COUNTIFS(aggregations!B:B,"*" & groups!C1& "*"))`

It takes more than 5 seconds to build the Hyperformula instance using the ‘buildFromSheets’ method. The expected aggregation sheet could have more than 20,000 records, so the application build time is long.

Adding a record, deleting multiple recods and update the cell value to the aggregation sheet freezes the application and takes around long time to complete the operation.

<img width="764" alt="Screenshot 2023-04-15 at 7 50 23 AM" src="https://user-images.githubusercontent.com/114735741/232178559-3b0cfb25-c3f3-4e16-ab94-8e0f926080eb.png">

I have created a demo on CodeSandbox and looking for a help :) 
[https://codesandbox.io/s/pivot-budget-aggregation-rx1kzt?file=/src/components/Container.tsx](https://codesandbox.io/s/pivot-budget-aggregation-rx1kzt?file=/src/components/Container.tsx)

**Comment by AMBudnik** [comment]
Hi @nuwanb-swivel 

If the following application structure is a must and tips from https://handsontable.com/docs/javascript-data-grid/performance/ (Performance tutorial) do not help much, it might be better to move the heaviest calculations to the server (Hyperformula also works on the backend). 
If you're interested, we can also conduct a code review to identify any potential areas of improvement in the current app. Feel free to contact us at support@handsontable.com to explore this option further.

**Reply by nuwanb-swivel** [reply]
Hi @AMBudnik 

Thank you for the tips. 

This structure is a must. We cannot change it. Also, the structure is a bit more complicated than on example I have provided. We have 3 levels of the parent-child relationship for sheets, Each sheet is connected to one another for calculations. As I explained, the expected data volume is more than 20000+ records. If we move calculations to the server side, we have to pass all the data back and forth when the user changes a single cell value, all these sheets need to recalculate. Considering the data volume I don't think the serverside calculation is an option here to improve the performance of the application.

One more thing, with the current structure and the data, the application works pretty well if we remove the above-explained formulas for calculating SUM and USAGE. Do you think is there anything or efficient formula we can use for calculations?

These sheets are in different tabs, Those are not always visible unless the user opens a tab. Is there a way we can suspend calculations for these columns if those tabs are not focused?

I will post real application screenshots here.

**Budget tab** This has 3 levels. When you click on a row it will drill down to a second level. When you click on the second level it will drill down to a third level.

**Frist Level and Second Level**

<img width="758" alt="Screenshot 2023-04-19 at 5 23 51 AM" src="https://user-images.githubusercontent.com/114735741/232929751-6fd10f8a-0d84-411a-9855-02638e89ffb8.png">

**Third Level**
<img width="1506" alt="Screenshot 2023-04-19 at 5 24 59 AM" src="https://user-images.githubusercontent.com/114735741/232930034-6abcdfb3-68eb-4192-9581-d11e83f7d5c1.png">

So inside the budget tabs, the user can navigate to L1 > L2 > L3

Other tabs are something similar to look-up data. All the codes are defined as namedexpersions. These namedexpressions codes will be used inside the L3 sheets on some columns for calculations.

Example: **CTD**. 

<img width="1512" alt="Screenshot 2023-04-19 at 5 27 21 AM" src="https://user-images.githubusercontent.com/114735741/232930210-ef31d157-f2ff-47fa-a49f-cca1bb208974.png">

We will try to contact handsontable support as well. :)

**Reply by AMBudnik** [reply]
Thank you for sharing the additional insights. We had a quick chat internally about this case and need some time to review it.

**Reply by nuwanb-swivel** [reply]
@AMBudnik Thank you very much.  Let us know if anything is needed from our side.

**Reply by AMBudnik** [reply]
Hi @nuwanb-swivel this task can take a little bit longer than expected. We just released Hyperformula v2.4.0, and need to solve one task before that. However, don't worry. Your task is on our priority list.

**Reply by nuwanb-swivel** [reply]
@AMBudnik thank you very much for the response.

**Reply by nilesh-swivel** [reply]
@AMBudnik any update regarding this?

**Reply by AMBudnik** [reply]
Hi @nilesh-swivel is your case the same as the one that the author reported? This task is scheduled for our developer to check but he's on a longer vacation (nearly a month) so all the tasks were postponed to release v2.5.0

**Comment by sequba** [comment]
Hi @nuwanb-swivel @nilesh-swivel,

The profiling of the provided example indicates that **~70%** of the build-time is spent on formula evaluation. The data contains **5000+** calls to the aggregation functions (`SUMIF`, `COUNTIF` & `COUNTIFS`), each of which searches linearly through **4000+** rows of the `aggregations` sheet. That is the cause of the long build-time.

### Recommended optimization

> The expected aggregation sheet could have more than 20,000 records

With this amount of data, one linear search is ok, but 4000 would be slow. The best solution is to reorganize your data to make the binary search possible. It will make a huge difference.

If the `aggregations` sheet was sorted by the most-searched column, you could:
- use [binary search](https://en.wikipedia.org/wiki/Binary_search_algorithm) to count occurrences by searching for lower and upper bound (requires writing a [custom function](https://hyperformula.handsontable.com/guide/custom-functions.html))
- calculate [prefix sums](https://en.wikipedia.org/wiki/Prefix_sum) and then use binary search to calculate total sums by searching for lower and upper bound (requires writing a [custom function](https://hyperformula.handsontable.com/guide/custom-functions.html))

### Other ideas for optimization

#### Reduce the number of heavy function calls

1. If there are function calls that calculate the same values, you can reuse the computed values by keeping them in the auxiliary cells or as [named expressions](https://hyperformula.handsontable.com/guide/named-expressions.html) or write a [custom function](https://hyperformula.handsontable.com/guide/custom-functions.html) using [memoization](https://en.wikipedia.org/wiki/Memoization).

2. In your code example, you have formulas like this:
```
COUNTIFS(aggregations!G:G,"*"&globals!C1&"*")+COUNTIFS(aggregations!H:H,"*"&globals!C1&"*")+COUNTIFS(aggregations!I:I,"*"&globals!C1&"*")+COUNTIFS(J:J,"*"&globals!C1&"*")
```
It can be simplified to:
```
COUNTIFS(aggregations!G:J,"*"&globals!C1&"*")
```

#### Optimize the creation of named expressions

You can add named expressions on build-time. This will be faster than adding them when the engine is already built.
```js
const namedExpressions = [
  { name: "A", expression: "=1+1" },
  { name: "B", expression: "=1+1+1" },
];
const hf = HyperFormula.buildFromSheets(sheets, options, namedExpressions);
```

#### Calculate the sums and counts more efficiently

You can create a [custom function](https://hyperformula.handsontable.com/guide/custom-functions.html) that will calculate the exact value that you need in a more efficient way. For example, calculate both count and sum on a single run or do something even more specific to your requirements.

#### Run HyperFormula server-side

Servers usually have more computation power than PCs, so the calculations will run faster.

> If we move calculations to the server side, we have to pass all the data back and forth when the user changes a single cell value, all these sheets need to recalculate. Considering the data volume I don't think the serverside calculation is an option here to improve the performance of the application.

It's possible to pass only the changed values after each operation, as HyperFormula [reports which cells have changed](https://hyperformula.handsontable.com/guide/basic-operations.html#what-are-the-changes). You don't need to send the full data every time.

#### Suspending calculations for selected sheet

> These sheets are in different tabs, Those are not always visible unless the user opens a tab. Is there a way we can suspend calculations for these columns if those tabs are not focused?

It is an interesting idea. HyperFormula does not have such a feature at the current stage of development, but it is an open-source project, so you can work on it yourself. I'll be happy to assist you ;)

I apologize for such a late reply. I hope this answer will be beneficial for making HyperFormula work for you. If you need help implementing these ideas or work on other solutions in greater detail, please [contact our sales team](https://hyperformula.handsontable.com/guide/contact.html#contact-sales) for information on our consulting services.

**Comment by nuwanb-swivel** [comment]
Hi @sequba Thanks for the detail explanation. I will try as you explained. I will post an update soon with the findings
